collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java

```diff
@@ -479,40 +487,71 @@ public void scrollTo(@Nullable QuestionWidget qw) {
-                            formController.saveAnswer(prompt.getIndex(),
-                                    ExternalAppsUtils.asStringData(bundle.get(key)));
-                            break;
-                        case Constants.DATATYPE_INTEGER:
-                            formController.saveAnswer(prompt.getIndex(),
-                                    ExternalAppsUtils.asIntegerData(bundle.get(key)));
-                            break;
-                        case Constants.DATATYPE_DECIMAL:
-                            formController.saveAnswer(prompt.getIndex(),
-                                    ExternalAppsUtils.asDecimalData(bundle.get(key)));
-                            break;
-                        default:
-                            throw new RuntimeException(
-                                    getContext().getString(R.string.ext_assign_value_error,
-                                            treeReference.toString(false)));
+
+        if (bundle != null) {
+            Set<String> keys = bundle.keySet();
+            for (String key : keys) {
+                for (QuestionWidget questionWidget : widgets) {
+                    FormEntryPrompt prompt = questionWidget.getFormEntryPrompt();
+                    TreeReference treeReference =
+                            (TreeReference) prompt.getFormElement().getBind().getReference();
+
+                    if (treeReference.getNameLast().equals(key)) {
+                        switch (prompt.getDataType()) {
+                            case Constants.DATATYPE_TEXT:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asStringData(bundle.get(key)));
+                                break;
+                            case Constants.DATATYPE_INTEGER:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asIntegerData(bundle.get(key)));
+                                break;
+                            case Constants.DATATYPE_DECIMAL:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asDecimalData(bundle.get(key)));
+                                break;
+                            default:
+                                throw new RuntimeException(
+                                        getContext().getString(R.string.ext_assign_value_error,
+                                                treeReference.toString(false)));
+                        }
+
+                        ((StringWidget) questionWidget).setDisplayValueFromModel();
+                        break;
                     }
+                }
+            }
+        }
+
+        if (clipData != null) {
+            for (int i = 0; i < clipData.getItemCount(); i++) {
+                CharSequence key = clipData.getItemAt(i).getText();
```

Comment:
See my comments on the test -- I don't think this matches the spec we had written up. It looks like it has the external app set both the text and a URI for each clip data item. The way we had written it up was to just set the URI and use an intent extra to do the match between Collect field name and clipdata URI. You should be able to just read the extra to match with the Collect field and then use the extra's value to get the file.

Reply:
The spec document says:
> Look for an intent extra with the name of the field. If one is found, attempt to exactly match the intent extra value to one of the ClipData items’ content URI
> If there’s a match, copy the file to the form instance folder and write its local filename in the corresponding form field

We don't really need to do that, I don't think. We can just try to resolve the URI from the extra as I described above. If that works, it means the ClipData was correctly populated.